### PR TITLE
Add InputStats to track bytes processed by a task

### PR DESCRIPTION
**Pull request by kfaraz**
This is based on #12750 and #10407

### Description
- Track bytes processed by a task and expose in task reports along with row stats
- Supported for classic batch and streaming ingestion
- Not supported for MSQ ingestion in this PR
- Not supported for `FirehoseToInputSourceReaderAdaptor`
- `processedBytes` measure the uncompressed input bytes. e.g. in case of the sample "wikipedia" datasource, it measures the total size of the unzipped `wikipedia.json` file (and not the `wikipedia.json.gz`)
- Records that are counted as `unparseable`, `thrownAway` or `processedWithErrors` also count towards `processedBytes`

### Implementation
- Add class `InputStats` to track processed bytes
- Add method `InputSourceReader.read(InputStats)` to read input rows while counting bytes.
> Since we need to count the bytes, we could not just have a wrapper around `InputSourceReader` or `InputEntityReader` (the way `CountableInputSourceReader` does) because the `InputSourceReader` only deals with `InputRow`s and the byte information is already lost.
- Classic batch: Use the new `InputSourceReader.read(inputStats)` in `AbstractBatchIndexTask`
- Streaming: Increment `processedBytes` in `StreamChunkParser`. This does not use the new `InputSourceReader.read(inputStats)` method.
- Extend `InputStats` with `RowIngestionMeters` so that bytes can be exposed in task reports

### Tests and refactors
- Update tests to verify the value of `processedBytes`
   - `HdfsInputSourceTest`
   - `S3InputSourceTest`
   - `GCSInputSourceTest`
   - `OssInputSourceTest`
   - `SqlInputSourceTest`
   - `DruidSegmentReaderTest`
   - `KinesisIndexTaskTest`
   - `KafkaIndexTaskTest`
- Rename `MutableRowIngestionMeters` to `SimpleRowIngestionMeters` and remove duplicate class
- Replace `CacheTestSegmentCacheManager` with `NoopSegmentCacheManager`
- Refactor `KafkaIndexTaskTest` and `KinesisIndexTaskTest`

### Pending items
- docs

### Further work
- Support for MSQ ingestion

### Release note
Track bytes processed by a task and publish them in the task report along with the row stats.

Sample row stats in a task report:
```json
"rowStats": {
    "buildSegments": {
        "processed": 24433,
        "processedBytes": 11956576,
        "processedWithError": 0,
        "thrownAway": 0,
        "unparseable": 0
    }
}
```

### Screenshots

Batch ingestion
image

Streaming ingestion
image

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: core/src/main/java/org/apache/druid/data/input/InputStats.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+{
+  default void incrementProcessedBytes(long incrementByValue)
+  {
+
+  }
+
+  default long getProcessedBytes()
+  {
+    return 0;
+  }
```
Comment (abhishekagarwal87) [inline core/src/main/java/org/apache/druid/data/input/InputStats.java:35]: why are default methods needed here?

Comment (kfaraz) [reply core/src/main/java/org/apache/druid/data/input/InputStats.java:35]: RowIngestionMeters is marked as an extension point, which is where this is used. I can move the default impls there.

Comment (kfaraz) [reply core/src/main/java/org/apache/druid/data/input/InputStats.java:35]: I am not sure if anyone uses their own impl of RowIngestionMeters. I would have preferred not having these default impls altogether.

Comment (kfaraz) [reply core/src/main/java/org/apache/druid/data/input/InputStats.java]: Moved to `RowIngestionMeters` for now.

File: core/src/main/java/org/apache/druid/data/input/InputSourceReader.java [hunk 2]
```diff
@@ -37,7 +37,12 @@
 @UnstableApi
 public interface InputSourceReader
 {
-  CloseableIterator<InputRow> read() throws IOException;
+  default CloseableIterator<InputRow> read() throws IOException
```
Comment (kfaraz) [inline core/src/main/java/org/apache/druid/data/input/InputSourceReader.java:40]: This method can now be removed as it is now only used in tests.

File: core/src/main/java/org/apache/druid/data/input/impl/FirehoseToInputSourceReaderAdaptor.java [hunk 3]
```diff
@@ -48,7 +49,7 @@ public FirehoseToInputSourceReaderAdaptor(
   }
 
   @Override
-  public CloseableIterator<InputRow> read() throws IOException
+  public CloseableIterator<InputRow> read(InputStats inputStats) throws IOException
```
Comment (kfaraz) [inline core/src/main/java/org/apache/druid/data/input/impl/FirehoseToInputSourceReaderAdaptor.java:52]: Maybe throw UnsupportedOpException here?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/external/ExternalInputSliceReader.java [hunk 4]
```diff
@@ -134,6 +136,7 @@ private static Iterator<SegmentWithDescriptor> inputSourceSegmentIterator(
     if (!temporaryDirectory.exists() && !temporaryDirectory.mkdir()) {
       throw new ISE("Cannot create temporary directory at [%s]", temporaryDirectory);
     }
+    final InputStats inputStats = new SimpleRowIngestionMeters();
```
Comment (kfaraz) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/external/ExternalInputSliceReader.java:139]: Fix this.

**Review by AmatyaAvadhanula — APPROVED** [review]
Thank you @kfaraz! LGTM, +1 after builds pass

**Comment by kfaraz** [comment]
Thanks a lot for the review, @AmatyaAvadhanula ! 
Thanks for doing the initial work on this, @somu-imply , @pjain1 !

**Pull request merged**